### ISSUE TITLE
Functions inside mut structs break the inliner (SSA refactor)

### ISSUE THREAD

**Issue by sirasistant** [noir-lang/noir#1844]
### Aim

Trying to have a struct to describe how to fetch notes from an oracle, one of the fields is a filter function. Having this struct with a function in it on a mutable variable makes function calls reach acir genereation:

```rust
struct FilterWrapper {
    skip: u8,
    filter: fn ([Field; 30], Field) -> [Field; 10],
}

fn filterImpl(fields: [Field; 30], default: Field) -> [Field; 10] {
    let mut result = [default; 10];
    for i in 0..5 {
        result[i] = fields[i + 5];
    }
    result
}

fn main(x: Field) -> pub [Field; 10] {
    let mut instance = FilterWrapper {
        skip: 0,
        filter: filterImpl ,
    };

    instance.skip = x as u8;

    let filter = instance.filter;
    filter([x; 30], 27)
}
```

### Expected Behavior

The example should compile

### Bug

```
% nargo compile t --experimental-ssa                                                                                                                                    ~/constructor mainframe
The application panicked (crashed).
Message:  internal error: entered unreachable code: expected calling a function
Location: crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs:261

This is a bug. We may have already fixed this in newer versions of Nargo so try searching for similar issues at https://github.com/noir-lang/noir/issues/.
If there isn't an open issue for this bug, consider opening one at https://github.com/noir-lang/noir/issues/new?labels=bug&template=bug_report.yml
```

This is the SSA inlining pass:
```
Initial SSA:
fn main f0 {
  b0(v0: Field):
    v3 = allocate
    store u8 0 at v3
    v4 = allocate
    store f1 at v4
    v5 = cast v0 as u8
    store v5 at v3
    v6 = load v4
    store v6 at v4
    v7 = load v3
    v8 = load v4
    v11 = call v8([v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0], Field 27)
    return v11
}
fn filterImpl f1 {
  b0(v0: [Field; 30], v1: Field):
    v3 = allocate
    store [v1, v1, v1, v1, v1, v1, v1, v1, v1, v1] at v3
    jmp b1(Field 0)
  b1(v4: Field):
    v7 = lt v4, Field 5
    jmpif v7 then: b2, else: b3
  b2():
    v8 = load v3
    v9 = add v4, Field 5
    v11 = array_get v0, index v9
    v12 = array_set v8, index v4, value v11
    store v12 at v3
    v13 = add v4, Field 1
    jmp b1(v13)
  b3():
    v14 = load v3
    return v14
}

After Inlining:
fn main f0 {
  b0(v0: Field):
    v3 = allocate
    store u8 0 at v3
    v4 = allocate
    store f1 at v4
    v5 = cast v0 as u8
    store v5 at v3
    v6 = load v4
    store v6 at v4
    v7 = load v3
    v8 = load v4
    v11 = call v8([v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0], Field 27)
    return v11
}
fn filterImpl f1 {
  b0(v0: [Field; 30], v1: Field):
    v3 = allocate
    store [v1, v1, v1, v1, v1, v1, v1, v1, v1, v1] at v3
    jmp b1(Field 0)
  b1(v4: Field):
    v7 = lt v4, Field 5
    jmpif v7 then: b2, else: b3
  b2():
    v8 = load v3
    v9 = add v4, Field 5
    v11 = array_get v0, index v9
    v12 = array_set v8, index v4, value v11
    store v12 at v3
    v13 = add v4, Field 1
    jmp b1(v13)
  b3():
    v14 = load v3
    return v14
}
fn main f2 {
  b0(v0: Field):
    v1 = allocate
    store u8 0 at v1
    v3 = allocate
    store f1 at v3
    v5 = cast v0 as u8
    store v5 at v1
    v6 = load v3
    store v6 at v3
    v7 = load v1
    v8 = load v3
    v11 = call v8([v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0], Field 27)
    return v11
}

```

### To Reproduce

1.
2.
3.
4.

### Installation Method

None

### Would you like to submit a PR for this Issue?

No

**Comment by jfecher** [comment]
This is a result of the `mem2reg` pass being after inlining/unrolling. We should likely add a call to `mem2reg` before inlining and verify it still works. Alternatively we could add some rudimentary per-block tracking of current values in the builder.

**Comment by jfecher** [comment]
Fixed by #1870
